operator/controllers/manifest_controller_test.go

```diff
@@ -42,60 +36,165 @@ func createManifest() *v1alpha1.Manifest {
+			Installs: []v1alpha1.InstallInfo{
+				{
+					Source: runtime.RawExtension{
+						Raw: specBytes,
+					},
+					Name: "oci-image",
+				},
+			},
+		})
+		Expect(k8sClient.Create(ctx, manifestObj)).Should(Succeed())
+		Eventually(getManifestState(client.ObjectKeyFromObject(manifestObj)), 5*time.Minute, 250*time.Millisecond).
+			Should(BeEquivalentTo(v1alpha1.ManifestStateError))
+		Expect(k8sClient.Delete(ctx, manifestObj)).Should(Succeed())
+		Eventually(getManifest(client.ObjectKeyFromObject(manifestObj)), 5*time.Minute, 250*time.Millisecond).
+			Should(BeTrue())
+		Expect(os.RemoveAll(util.GetFsChartPath(imageSpec))).Should(Succeed())
+		return true
+	}
+}
+
+func getManifestState(key client.ObjectKey) func() v1alpha1.ManifestState {
+	return func() v1alpha1.ManifestState {
 		manifest := v1alpha1.Manifest{}
 		err := k8sClient.Get(ctx, key, &manifest)
 		if err != nil {
 			return ""
 		}
-		return string(manifest.Status.State)
+		return manifest.Status.State
 	}
 }
 
-func createRepo() error {
-	err := os.MkdirAll(helmCacheHome, os.ModePerm)
-	if err != nil {
-		return err
+func getManifest(key client.ObjectKey) func() bool {
+	return func() bool {
+		manifest := v1alpha1.Manifest{}
+		err := k8sClient.Get(ctx, key, &manifest)
+		return errors.IsNotFound(err)
 	}
+}
+
+func setHelmEnv() error {
 	os.Setenv(helmCacheHomeEnv, helmCacheHome)
 	os.Setenv(helmCacheRepoEnv, helmCacheRepo)
+	os.Setenv(helmRepoEnv, helmRepoFile)
 	return nil
 }
 
-func deleteRepo() error {
-	err := os.RemoveAll(helmCacheHome)
-	if err != nil {
-		return err
-	}
+func unsetHelmEnv() error {
 	os.Unsetenv(helmCacheHomeEnv)
 	os.Unsetenv(helmCacheRepoEnv)
+	os.Unsetenv(helmRepoEnv)
 	return nil
 }
 
-var _ = Describe("given manifest with", Ordered, func() {
-	manifest := createManifest()
+var _ = Describe("given manifest with a helm repo", Ordered, func() {
 	BeforeAll(func() {
-		Expect(createRepo()).Should(Succeed())
-	})
-	BeforeEach(func() {
-		Expect(k8sClient.Create(ctx, manifest)).Should(Succeed())
-	})
-	AfterEach(func() {
-		Expect(k8sClient.Delete(ctx, manifest)).Should(Succeed())
+		Expect(setHelmEnv()).Should(Succeed())
 	})
 
-	It("Should result in a ready state", func() {
-		By("having transitioned the CR State to Ready")
-		Eventually(getManifestState(client.ObjectKeyFromObject(manifest)), 5*time.Minute, 250*time.Millisecond).
-			Should(BeEquivalentTo(string(v1alpha1.ManifestStateReady)))
-	})
+	DescribeTable("given watcherCR reconcile loop",
+		func(testCaseFn func() bool) {
+			Expect(testCaseFn()).To(BeTrue())
+		},
+		[]TableEntry{
+			Entry("when manifestCR contains a valid helm repo", createManifestWithHelmRepo()),
+			Entry("when manifestCRs contain valid OCI Image specification", createManifestWithOCI()),
+			Entry("when manifestCR contains invalid OCI Image specification", createManifestWithInvalidOCI()),
+		})
 
 	AfterAll(func() {
-		Expect(deleteRepo()).Should(Succeed())
+		Expect(unsetHelmEnv()).Should(Succeed())
```

Comment:
this you can put in the beginning of the test so its clear from the start what will be cleaned up

Reply:
`setHelmEnv` is called in `BeforeAll` which should override all necessary env variables, or do you mean something else?

Reply:
just that you should write your AfterAll always in the beginning of a test, not at the end. Still will work fine since its async :)